### ISSUE TITLE
Helm upgrade doesn't uninstall old resource properly

### ISSUE THREAD

**Issue by suxiaoxiaomm** [helm/helm#12085]
Hi Dear Experts,
I am currently trying to install a Knative Service via helm charts.

I want to use a helm value(as below "knativeService") to control whether to deploy a Knative Service or a normal deployment + normal service.
When the "knativeService" is equal to "true", it will deploy knative Service, otherwise, it deploy a normal deployment + normal service.

However, during the switch( I am using the command: **helm upgrade -i** ), it will stuck at uninstall knativeService.

It seems when using Knative Service, there is an ExternalName type k8s service will be created. When I switch to a normal k8s deployment, it will also try to create a k8s service which has the same name as previous ExternalName service.

This caused the Knative controller confused and reporting error. And thus, the Knative Service CR cannot be uninstalled successfully: Similarly, If I want to switch from normal k8s deployment to Knative Service, still go problems.

But I can do helm uninstall first, then try to do install. This would properly switch.

**So I am wondering:**
**1. What is the mechanism for using "helm upgrade"? Does it uninstall unnecessary old resource first, then try to install the new resource?** 
**2. Is there a way to guarantee the upgrade workflow's order to make sure the unnecessary resources got delete first?**

Appreciate any suggestions!

Below is my helm template:

```
---
{{ if eq .Values.knativeService "true" }}
apiVersion: serving.knative.dev/v1
kind: Service
metadata:
  name: {{ .Values.serviceName }}
  namespace: {{ .Values.namespace }}
spec:
  template:
    metadata:
      annotations:
        autoscaling.knative.dev/maxScale: "5"
    spec:
      containers:
      - name: {{ $container }}
        image: {{ $registry }}/{{ $container }}:{{ (tuple $scope ( index $imageVersionsList $i)) | include "imageVersion" }}

{{ else }}
---
apiVersion: v1
kind: Service
metadata:
  name: {{ .Values.serviceName }}
spec:
  selector:
    app: {{ .Values.serviceName }}
  ports:
    - name: http-web
      port: 80
      targetPort: {{ default "80" .Values.servicePort }}
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: {{ .Values.serviceName }}
  namespace: {{ .Values.namespace }}
spec:
  replicas: 1
  selector:
    matchLabels:
      app: {{ .Values.serviceName }}
  template:
    metadata:
      labels:
        app: {{ .Values.serviceName }}
    spec:
      terminationGracePeriodSeconds: 10
      containers:
      - name: {{ $container }}
        image: {{ $registry }}/{{ $container }}:{{ (tuple $scope ( index $imageVersionsList $i)) | include "imageVersion" }}
        imagePullPolicy: "Always"

{{ end }}

```

**Comment by joejulian** [comment]
Helm will uninstall a resource that was installed before (exists in the current release manifest) but is not in the new release manifest. Otherwise, if the resource exists it will apply a patch to alter the resource to match the new desired state, similar to `kubectl apply`. It doesn't delete or apply first, it walks through the resources and affects each individually in turn:
https://github.com/helm/helm/blob/2398830f183b6d569224ae693ae9215fed5d1372/pkg/kube/client.go#L433-L454

The only idea I can think of would be to have a pre-upgrade hook that runs a kubectl command to pre-delete the Service that's being replaced.

**Comment by suxiaoxiaomm** [comment]
@joejulian Thanks a lot for the suggestion. It is clear now. Let me take some look about this pre-upgrade hook or just use a shell script directly before conduct the upgrade.

**Comment by tiandingyi** [comment]
I had a similar problem.
The hpa version I used was autoscaling/v2beta1 ,and helm version is 3.6.3
And then I upgrade helm to 3.10.3.
when I use helm upgrade again,  I got this error

Error: UPGRADE FAILED: unable to build kubernetes objects from current release manifest: resource mapping not found for name: "*********" namespace: "" from "": no matches for kind "HorizontalPodAutoscaler" in version "autoscaling/v2beta1"
 
So I changed the version of hpa to autoscaling/v2，I still got the same error. I have to use old helm uninstall old helm chart，because I will be refused by new version helm when I want to delete old helm resource
